IPSec key rotation without agent restart

jibi opened the pull request.

jibi committed:
nodediscovery: split StartDiscovery into smaller methods

in preparation to a subsequent change, this commit moves some of the
(*NodeDiscovery).StartDiscovery() logic into separated methods:

* fillLocalNode(), responsible for syncing the localNode object with the
  actual state of the node
* updateLocalNode(), responsible for publishing the updated KV store
  entry and/or CiliumNode object for the local node

In addition to that, this commit introduces also the UpdateLocalNode()
method, which ensures that the proper lock is held while syncing the
localNode object state with fillLocalNode() and then publishing the
new object with updateLocalNode().

No functional changes are introduced.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

christarazi commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -620,3 +625,49 @@ func DeleteIPsecEncryptRoute() {
+
+func keyfileWatcher(watcher *fsnotify.Watcher, keyfilePath string, nodediscovery *nodediscovery.NodeDiscovery) {
+	for {
+		select {
```
Would be good to wire up daemon's `context` for this loop to exit on shutdown

kaworu commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -179,6 +179,46 @@ func (n *NodeDiscovery) StartDiscovery() {
+	go func() {
+		select {
+		case <-n.Registered:
+		case <-time.NewTimer(defaults.NodeInitTimeout).C:
```
nit
```suggestion
		case <-time.After(defaults.NodeInitTimeout):
```

kaworu reviewed (commented):
@jibi FWIW we have some wrapper and logic around fsnotify to watch over TLS certificates [here](https://github.com/cilium/cilium/tree/master/pkg/crypto/certloader). The reason I'm mentioning this is that watching over k8s secrets has proven to be tricky:

1. Because of the symlinks structure and replacement hapenning when the "underlying" secret change
2. Because we get "a lot of events" (i.e. more than one) when a secret change
3. Because certloader handle `optional` volumes (not the case for ipsec keys if I understand correctly)

For 1. a [`fswatcher.Watcher`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/fswatcher/fswatcher.go) could be re-used here. For 2. we could take inspiration of certloader's [event coalescing mechanism](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/watcher.go#L132-L133).

gandro commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -620,3 +625,49 @@ func DeleteIPsecEncryptRoute() {
+}
+
+func StartKeyfileWatcher(keyfilePath string, nodediscovery *nodediscovery.NodeDiscovery) error {
+	watcher, err := fsnotify.NewWatcher()
```
Using fsnotify to watch for Kubernetes secret changes might not always work, due to it using symlinks (see #13621 for how to reproduce). Have you checked if you could use the `fswatcher` package instead?

https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/fswatcher/fswatcher.go#L25-L50

Edit: @kaworu beat me to it.

jibi committed:
Move fswatcher package out of crypto/certloader

Given that this logic is totally agnostic to the crypto/certloader
package and in a subsequent commit it will be used also in the IPSec
subsystem, make it a top level package.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: use time.After() instead of time.NewTimer().C

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -284,6 +292,17 @@ func (n *NodeDiscovery) StartDiscovery() {
+	}
+
+	n.updateLocalNode()
+}
```
I don't see any reference to this function in the commit description? Wrong commit maybe?

jibi replied:
Forgot to update the commit message

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -620,3 +626,44 @@ func DeleteIPsecEncryptRoute() {
+			nodediscovery.UpdateLocalNode(true)
+
+		case err := <-watcher.Errors:
+			log.WithError(err).WithField("path", keyfilePath).Warning("error encountered while watching file with fsnotify")
```
Nit:
```suggestion
			log.WithError(err).WithField(logfields.Path, keyfilePath).
 				Warning("error encountered while watching file with fsnotify")
```

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -620,3 +626,44 @@ func DeleteIPsecEncryptRoute() {
+				continue
+			}
+
+			log.WithField(logfields.Path, keyfilePath).Debugf("Received %s fsnotify event for IPSec keyfile", event.Op.String())
```
Should we make this an Info message? I'm not sure if we have other log messages for when the secret changes.

jibi replied:
We actually have already this information logged in the `fswatcher` package:
```
level=debug msg="Received fsnotify event" file-path=/etc/ipsec/keys operation=CHMOD subsys=fswatcher
```

I think I'll get rid of this and add an `Info` logline to `LoadIPSecKeysFile()` so that we log events only when the keys get actually rotated

pchaigno reviewed (approved):
Only minor comments. Nothing blocking.

Thanks for the nice commit history! :pray:

kaworu reviewed (changes requested):
Thanks @jibi, overall LGTM. Now an assumption made before this patch about `ipSecKeysGobal` being safe to read without locking don't hold anymore: https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L45-L47

So I believe we have to:

1. add a `sync.RWMutex` protecting [`ipSecKeysGlobal`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L45-L47)
2. update the comment
3. read-lock on [`getIPSecKeys`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L49)
4. write-lock on [`loadIPSecKeys`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L481)

I'm _guessing_ writing without locking in `pkg/datapath/linux/ipsec/ipsec_linux_test.go` is still ok though, but to be checked.

kaworu commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -620,3 +628,40 @@ func DeleteIPsecEncryptRoute() {
+			log.WithError(err).WithField(logfields.Path, keyfilePath).Warning("Error encountered while watching file with fsnotify")
+
+		case <-ctx.Done():
+			return
```
```suggestion
		case <-ctx.Done():
			watcher.Stop()
			return
```

kaworu commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -42,11 +49,18 @@ type ipSecKey struct {
 
 func getIPSecKeys(ip net.IP) *ipSecKey {
+	ipSecKeysGlobalLock.RLock()
+	defer ipSecKeysGlobalLock.RLock()
```
```suggestion
	defer ipSecKeysGlobalLock.RUnlock()
```

jibi replied:
:man_facepalming:

kaworu commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -483,6 +499,9 @@ func loadIPSecKeys(r io.Reader) (int, uint8, error) {
 	scopedLog := log
 
+	ipSecKeysGlobalLock.Lock()
+	defer ipSecKeysGlobalLock.Lock()
```
```suggestion
	defer ipSecKeysGlobalLock.Unlock()
```

jibi committed:
ipsec: implement keyfile watcher

This commit adds a new filesystem watcher for the IPSec subsystem, which
in turn will enable automatic key rotation without the need to restart
the agent whenever a change to the IPSec keyfile is detected.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

ldelossa reviewed (approved):
LGTM!
